eeps/eep-0058.md

```diff
@@ -0,0 +1,255 @@
+
+`MapOrIterator` is any Erlang expression that evaluates to `map()`
+or `maps:iterator()` datatype.
+
+### Semantics
+
+Map comprehension should be semantically equivalent to the
+`maps:from_list/1` call over a list comprehension producing
+a list of 2-tuples:
+
+    #{K => V || ...}
+
+is equivalent to
+
+    maps:from_list([{K, V} || ...])
+
+
+Map generator should be semantically equivalent to the list generator
+running over the result of `maps:to_list/1`, when the input is a
+map datatype:
+
+    #{ ... || K := V <- Map}
+
+is equivalent to
+
+    #{ ... || {K, V} <- maps:to_list(Map)}
+
+And for `maps:iterator()` as input - it can't be easily expressed as
+equivalent list generator. But the idea is that it's equivalent to
+the consumption of the iterator by repetitive calls to `maps:next/1`
+and matching `Key` and `Value` over key and value patterns
+accordingly on each iteration loop.
+
+Map generator should raise `error({bad_generator, ExprResult})`
+if the `MapOrIterator` expression evaluation produced anything but
+map or map iterator.
+
+Key pattern is allowed to be catch-all / contain unbound variables
+(which is not allowed in other map pattern-matching contexts).
+
+Variable binding and shadowing rules should be the same as for list
+and binary comprehensions (which follows from the equivalence rules
+described above).
+
+Rationale
+=========
+
+### Syntax for arrow, generator pattern and comprehension key-value pair
+
+We decided to go with the syntax proposed in EEP-43.
+
+`#{ .. }` were chosen for the outer tokens of map comprehension
+because the same are used when constructing a new map - rule similar to
+list and binary comprehensions.
+
+`K := V <-` is not ambiguous for the parser, so unlike binary
+generator, there is no need for a new arrow. `<-` was chosen over
+`<=` because the latter is less visually distinguishable when it is
+placed relatively close to the `=>` and `:=` tokens:
+
+    #{K => V || K := V <= Map}.
+    % vs
+    #{K => V || K := V <- Map}.
+
+`KeyPattern := ValuePattern` is the same syntax that is used in
+map pattern-matching (however, in map generators KeyPattern is also
+allowed to be a match-all / unbound variable).
+
+`KeyExpr => ValueExpr` is the same syntax that is used for the new
+map construction.
+
+### Order in which map generator yields key-value pairs
+
+Since maps do not have any order specified, the order in which the
+generator would yield key-value pairs should not be guaranteed (same
+way as the ordering of the list, resulting from `maps:to_list/1` is
+not guaranteed).
+
+### Behavior when comprehension produced duplicate keys
+
+Same as `maps:from_list/1`: the latest (right-most) value is used
+and the previous values are ignored. It might be less predictable
+when a map comprehension is used in combination with a map generator
+(because the map generator order is not defined), but is important when
+a map comprehension is used with a list or a binary generator.
```

Comment:
This paragraph is a bit confusing I think, maybe you should phrase it differently?

To be sure, I get the idea: If a map generator is used to generate key-value pairs, and the comprehension produces possibly duplicate keys based on them, it is not predictable which value will be in the resulting map for that key. For example, if a map comprehension is used to flip keys and values in a map, with...
```erlang
#{ V => K || K := V <- #{a => x, b => x} }
```
... the result could be either `#{x => a}` or `#{x => b}`, which one is not predictable (or guaranteed).

With a list (or binary) generator instead, the result is predictable. With...
```erlang
#{ V => K || {K, V} <- [{a, x}, {b, x}] }
```
... the result will _always_ be `#{x => b}`, and with...
```erlang
#{ V => K || {K, V} <- [{b, x}, {a, x}] }
```
... the result will _always_ be `#{x => a}`.

Reply:
Yes, this is exactly what I tried to say.

Reply:
"right-most" isn't really true here. Latest yes. Last generated maybe.

Reply:
I actually copy-pasted this phrase "the latest (right-most) value is used and the previous values are ignored" from OTP documentation of `maps:from_list` http://erlang.org/doc/man/maps.html#from_list-1